Export the public classes, functions and explicit template
instantiations for the component build, with SHELL_ENCRYPTION_EXPORT
from base/shell_encryption_export.h.

Chromium commit f2bcadd8fbfe7.

diff --git a/src/dft_transformations_hwy.h b/src/dft_transformations_hwy.h
index 174d43ffc9c7b..20f7458419512 100644
--- a/src/dft_transformations_hwy.h
+++ b/src/dft_transformations_hwy.h
@@ -22,6 +22,7 @@
 #include "absl/status/status.h"
 #include "ntt_parameters.h"
 #include "status_macros.h"
+#include "third_party/shell-encryption/base/shell_encryption_export.h"
 
 namespace rlwe::internal {
 
@@ -29,7 +30,7 @@ namespace rlwe::internal {
 // in-place, where q is a prime and N is a power of two such q == 1 (mod 2N).
 // The input polynomial is represented by its coefficient vector `coeffs`.
 template <typename ModularInt>
-extern absl::Status ForwardNumberTheoreticTransformHwy(
+extern SHELL_ENCRYPTION_EXPORT absl::Status ForwardNumberTheoreticTransformHwy(
     std::vector<ModularInt>& coeffs,
     const NttParameters<ModularInt>& ntt_params,
     const typename ModularInt::Params& mod_params);
diff --git a/src/int256.h b/src/int256.h
index 27e6fcc63e11d..2b0f2cd0df3bc 100644
--- a/src/int256.h
+++ b/src/int256.h
@@ -18,13 +18,14 @@
 
 #include "absl/numeric/int128.h"
 #include "integral_types.h"
+#include "third_party/shell-encryption/base/shell_encryption_export.h"
 
 namespace rlwe {
 
 struct uint256_pod;
 
 // An unsigned 256-bit integer type. Thread-compatible.
-class uint256 {
+class SHELL_ENCRYPTION_EXPORT uint256 {
  public:
   constexpr uint256();
   constexpr uint256(absl::uint128 top, absl::uint128 bottom);
@@ -110,7 +111,7 @@ class uint256 {
 #endif
 
   // We add "std::" to avoid including all of port.h.
-  friend std::ostream& operator<<(std::ostream& o,
+  friend SHELL_ENCRYPTION_EXPORT std::ostream& operator<<(std::ostream& o,
                                                           const uint256& b);
 
  private:
@@ -169,7 +170,8 @@ const uint256_pod kuint256max = {absl::Uint128Max(), absl::Uint128Max()};
 #endif
 
 // allow uint256 to be logged
-extern std::ostream& operator<<(std::ostream& o, const uint256& b);
+extern SHELL_ENCRYPTION_EXPORT std::ostream& operator<<(std::ostream& o,
+                                                  const uint256& b);
 
 // Methods to access low and high pieces of 256-bit value.
 // Defined externally from uint256 to facilitate conversion
diff --git a/src/montgomery.cc b/src/montgomery.cc
index a064cd32d7a68..c9f8868e474a3 100644
--- a/src/montgomery.cc
+++ b/src/montgomery.cc
@@ -537,17 +537,17 @@ absl::StatusOr<MontgomeryInt<T>> MontgomeryInt<T>::MultiplicativeInverseFast(
 
 // Instantiations of MontgomeryInt and MontgomeryIntParams with specific
 // integral types.
-template struct MontgomeryIntParams<Uint16>;
-template struct MontgomeryIntParams<Uint32>;
-template struct MontgomeryIntParams<Uint64>;
-template struct MontgomeryIntParams<absl::uint128>;
-template class MontgomeryInt<Uint16>;
-template class MontgomeryInt<Uint32>;
-template class MontgomeryInt<Uint64>;
-template class MontgomeryInt<absl::uint128>;
+template struct EXPORT_TEMPLATE_DEFINE(SHELL_ENCRYPTION_EXPORT) MontgomeryIntParams<Uint16>;
+template struct EXPORT_TEMPLATE_DEFINE(SHELL_ENCRYPTION_EXPORT) MontgomeryIntParams<Uint32>;
+template struct EXPORT_TEMPLATE_DEFINE(SHELL_ENCRYPTION_EXPORT) MontgomeryIntParams<Uint64>;
+template struct EXPORT_TEMPLATE_DEFINE(SHELL_ENCRYPTION_EXPORT) MontgomeryIntParams<absl::uint128>;
+template class EXPORT_TEMPLATE_DEFINE(SHELL_ENCRYPTION_EXPORT) MontgomeryInt<Uint16>;
+template class EXPORT_TEMPLATE_DEFINE(SHELL_ENCRYPTION_EXPORT) MontgomeryInt<Uint32>;
+template class EXPORT_TEMPLATE_DEFINE(SHELL_ENCRYPTION_EXPORT) MontgomeryInt<Uint64>;
+template class EXPORT_TEMPLATE_DEFINE(SHELL_ENCRYPTION_EXPORT) MontgomeryInt<absl::uint128>;
 
 #ifdef ABSL_HAVE_INTRINSIC_INT128
-template struct MontgomeryIntParams<unsigned __int128>;
-template class MontgomeryInt<unsigned __int128>;
+template struct EXPORT_TEMPLATE_DEFINE(SHELL_ENCRYPTION_EXPORT) MontgomeryIntParams<unsigned __int128>;
+template class EXPORT_TEMPLATE_DEFINE(SHELL_ENCRYPTION_EXPORT) MontgomeryInt<unsigned __int128>;
 #endif
 }  // namespace rlwe
diff --git a/src/montgomery.h b/src/montgomery.h
index 593c13586407a..eafde7ff1c097 100644
--- a/src/montgomery.h
+++ b/src/montgomery.h
@@ -34,6 +34,8 @@
 #include "serialization.pb.h"
 #include "status_macros.h"
 #include "statusor.h"
+#include "third_party/shell-encryption/base/shell_encryption_export.h"
+#include "third_party/shell-encryption/base/shell_encryption_export_template.h"
 
 namespace rlwe {
 
@@ -667,18 +669,18 @@ class ABSL_MUST_USE_RESULT MontgomeryInt {
 
 // Instantiations of MontgomeryInt and MontgomeryIntParams with specific
 // integral types.
-extern template struct MontgomeryIntParams<Uint16>;
-extern template struct MontgomeryIntParams<Uint32>;
-extern template struct MontgomeryIntParams<Uint64>;
-extern template struct MontgomeryIntParams<absl::uint128>;
-extern template class MontgomeryInt<Uint16>;
-extern template class MontgomeryInt<Uint32>;
-extern template class MontgomeryInt<Uint64>;
-extern template class MontgomeryInt<absl::uint128>;
+extern template struct EXPORT_TEMPLATE_DECLARE(SHELL_ENCRYPTION_EXPORT) MontgomeryIntParams<Uint16>;
+extern template struct EXPORT_TEMPLATE_DECLARE(SHELL_ENCRYPTION_EXPORT) MontgomeryIntParams<Uint32>;
+extern template struct EXPORT_TEMPLATE_DECLARE(SHELL_ENCRYPTION_EXPORT) MontgomeryIntParams<Uint64>;
+extern template struct EXPORT_TEMPLATE_DECLARE(SHELL_ENCRYPTION_EXPORT) MontgomeryIntParams<absl::uint128>;
+extern template class EXPORT_TEMPLATE_DECLARE(SHELL_ENCRYPTION_EXPORT) MontgomeryInt<Uint16>;
+extern template class EXPORT_TEMPLATE_DECLARE(SHELL_ENCRYPTION_EXPORT) MontgomeryInt<Uint32>;
+extern template class EXPORT_TEMPLATE_DECLARE(SHELL_ENCRYPTION_EXPORT) MontgomeryInt<Uint64>;
+extern template class EXPORT_TEMPLATE_DECLARE(SHELL_ENCRYPTION_EXPORT) MontgomeryInt<absl::uint128>;
 
 #ifdef ABSL_HAVE_INTRINSIC_INT128
-extern template struct MontgomeryIntParams<unsigned __int128>;
-extern template class MontgomeryInt<unsigned __int128>;
+extern template struct EXPORT_TEMPLATE_DECLARE(SHELL_ENCRYPTION_EXPORT) MontgomeryIntParams<unsigned __int128>;
+extern template class EXPORT_TEMPLATE_DECLARE(SHELL_ENCRYPTION_EXPORT) MontgomeryInt<unsigned __int128>;
 #endif
 }  // namespace rlwe
 
diff --git a/src/ntt_parameters.h b/src/ntt_parameters.h
index 022a6eb42fc42..79540d06cc159 100644
--- a/src/ntt_parameters.h
+++ b/src/ntt_parameters.h
@@ -24,6 +24,7 @@
 #include "constants.h"
 #include "status_macros.h"
 #include "statusor.h"
+#include "third_party/shell-encryption/base/shell_encryption_export.h"
 
 namespace rlwe {
 namespace internal {
@@ -93,7 +94,7 @@ rlwe::StatusOr<std::vector<ModularInt>> NttPsis(
 // Creates a vector containing the indices necessary to perform the NTT bit
 // reversal operation. Index i of the returned vector contains an integer with
 // the rightmost log_n bits of i reversed.
-std::vector<unsigned int> BitrevArray(
+SHELL_ENCRYPTION_EXPORT std::vector<unsigned int> BitrevArray(
     unsigned int log_n);
 
 // Helper function: Perform the bit-reversal operation in-place on coeffs_.
diff --git a/src/prng/chacha_prng.h b/src/prng/chacha_prng.h
index d1584468130f0..d12f131e169c7 100644
--- a/src/prng/chacha_prng.h
+++ b/src/prng/chacha_prng.h
@@ -26,10 +26,11 @@
 #include "prng/chacha_prng_util.h"
 #include "prng/prng.h"
 #include "statusor.h"
+#include "third_party/shell-encryption/base/shell_encryption_export.h"
 
 namespace rlwe {
 
-class ChaChaPrng : public SecurePrng {
+class SHELL_ENCRYPTION_EXPORT ChaChaPrng : public SecurePrng {
  public:
   // Constructs a secure pseudorandom number generator using the ChaCha20 stream
   // cipher. The parameter in_key is the key for the ChaCha20.
diff --git a/src/prng/chacha_prng_util.h b/src/prng/chacha_prng_util.h
index 2feb550d3b1f3..dfc996ceecca0 100644
--- a/src/prng/chacha_prng_util.h
+++ b/src/prng/chacha_prng_util.h
@@ -27,6 +27,7 @@
 #include "absl/strings/string_view.h"
 #include "integral_types.h"
 #include "statusor.h"
+#include "third_party/shell-encryption/base/shell_encryption_export.h"
 
 namespace rlwe {
 namespace internal {
@@ -37,24 +38,24 @@ const int kChaChaOutputBytes = 255 * 32;
 
 // Once pseudorandom output is exhausted, the salt is updated to construct
 // new pseudorandom output.
-absl::Status ChaChaPrngResalt(
+SHELL_ENCRYPTION_EXPORT absl::Status ChaChaPrngResalt(
     absl::string_view key, int buffer_size, int* salt_counter,
     int* position_in_buffer, std::vector<Uint8>* buffer);
 
 // Generates a secure key for instantiating an CHACHA.
-rlwe::StatusOr<std::string> ChaChaPrngGenerateKey();
+SHELL_ENCRYPTION_EXPORT rlwe::StatusOr<std::string> ChaChaPrngGenerateKey();
 
 // Returns 8 bits of randomness.
 //
 // Fails on internal cryptographic errors.
-rlwe::StatusOr<Uint8> ChaChaPrngRand8(
+SHELL_ENCRYPTION_EXPORT rlwe::StatusOr<Uint8> ChaChaPrngRand8(
     absl::string_view key, int* position_in_buffer, int* salt_counter,
     std::vector<Uint8>* buffer);
 
 // Returns 64 bits of randomness.
 //
 // Fails on internal cryptographic errors.
-rlwe::StatusOr<Uint64> ChaChaPrngRand64(
+SHELL_ENCRYPTION_EXPORT rlwe::StatusOr<Uint64> ChaChaPrngRand64(
     absl::string_view key, int* position_in_buffer, int* salt_counter,
     std::vector<Uint8>* buffer);
 
diff --git a/src/prng/single_thread_chacha_prng.h b/src/prng/single_thread_chacha_prng.h
index 609b7ab749767..449e5f8f6d08f 100644
--- a/src/prng/single_thread_chacha_prng.h
+++ b/src/prng/single_thread_chacha_prng.h
@@ -24,10 +24,11 @@
 #include "prng/chacha_prng_util.h"
 #include "prng/prng.h"
 #include "statusor.h"
+#include "third_party/shell-encryption/base/shell_encryption_export.h"
 
 namespace rlwe {
 
-class SingleThreadChaChaPrng : public SecurePrng {
+class SHELL_ENCRYPTION_EXPORT SingleThreadChaChaPrng : public SecurePrng {
  public:
   // Constructs a secure pseudorandom number generator using the ChaCha20 stream
   // cipher. The parameter in_key is the key for the ChaCha20.
